### ISSUE TITLE
Cannot find definitions in remote.containers.repositoryConfigurationPaths when building a devcontainer

### ISSUE THREAD

**Issue by kalfa** [microsoft/vscode-remote-release#6013]
# intro

There are cryptic bugs and feature requests about `remote.containers.repositoryConfigurationPaths`, but nothing really explains how to use it.

https://github.com/microsoft/vscode-dev-containers/tree/main/repository-containers says pretty nothing, beside what I can interpret as "there is a possibility to not include in your repository the .devcontainer definition, but keep it outside. by default a specific repository owned by Microsoft is checkd, but users can use their own by adding it to remote.containers.repositoryConfigurationPaths".

It's not clear what should really go in that setting. though. Some bugs shows a filesystem path which is actually a git repository, e.g. #3310 and related, but I cannot make it work.

What is the structure of the repository?

# what I am trying to do

I have a gitlab.com (not github!) group and some repos, for which I want to lookup devcontainers definitions, and not having them inline in each repository.

So I created another repository with the following structure

```
$ ls -R
README.md             repository-containers

./repository-containers:
gitlab.some.hosted.instance

./repository-containers/gitlab.some.hosted.instance:
group

./repository-containers/gitlab.some.hosted.instance/group:
repository1     repository1.git

/repository-containers/gitlab.some.hosted.instance/group/repository1:
.devcontainer

./repository-containers/gitlab.some.hosted.instance/group/repository1/.devcontainer:
Dockerfile             devcontainer.json      setup_dev_container.sh otherdata/
```

The devcontainer definitions works if placed in the workspace/project repository.

I duplicated the definition with and without the .git extension in case it was the problem. it is not.

Then I set
```
"remote.containers.repositoryConfigurationPaths": [
  "/Users/kalfa/Sources/devcontainers-repository/",
  "/Users/kalfa/Sources/devcontainers-repository/repository-containers/",
  "git@gitlab.some.hosted.instance/group/devcontainers-repository.git",
  "git@gitlab.some.hosted.instance/group/devcontainers-repository.git/repository-containers/"
]
```

I know that the "git@gitlab...." works because I use a similar url for dotfiles, I would expect to be usable as well there, unless of coruse URIs are not meant (in such case an error would be appreciated).

From some bug reports, it seems that I can only put filesystem path in the array, so I placed also the path where I cloned the same repository.

I added 'repository-containers' in case it requires it, to some entries.

Then I build the devcontainer in the currently open workspace with "rebuild and reopen in container".

Also I tried, "Add devlopment container configuration files ..." which still does not find the match in the array, in case the settings was to lookup and a definition locally.

# Current behaviour

When I build the devcotainer and no .devcontainer is present, a menu for selecting a container configuraiton definition appears, none of the options match any item of the array.

No sign the configuration is erroring, or ignored (how can I debug it?)

# Expected behaviour

When I build the devcontainer for a repository with "rebuild and reopen in container", since I dont' have a `.devcontainer` directory in the workspace, the settings will be used instead:
At least one of the items are matched (first come first served) or an error is clearly shown that it does not make sense what I placed there.
In fact if I place non existing directories or URIs it does not complain.

Also, it would help a lot a complete documentation of how the setting works and how to explit it. 
https://github.com/microsoft/vscode-dev-containers and https://github.com/microsoft/vscode-dev-containers/tree/main/repository-containers do not explain it, they try to give a vague gist of what it is about.

**Comment by chrmarti** [comment]
What is the output of `git remote -v` in your project repository?

**Comment by kalfa** [comment]
Hi @chrmarti , I think you are on the right path. the remote seems to be set to an alias of the official FQDN for some reason.
I will try to ASAP and let you know if that resolved the problem.

```shell
$ git remote -v
gitlab
origin	git@gitlab.some.hosted.instance:group/repository1.git (fetch)
origin	git@gitlab.some.hosted.instance:group/repository1.git (push)
```

(redacted of course)

Can this information be written explicitly somewhere, to help people understand the mechanism and where possible mistakes are made? I'm sure this would boost the understanding and use of this feature!

thanks

**Comment by kalfa** [comment]
Apologies for the delayed answer.

Yes, it works.

Is it possible to add clearly in one of the pages I mentioned, how it works?

There is lots of guess works to do, and it's a really useful feature which should be more accessible

Thanks!

**Comment by marcussharpe** [comment]
I'm having a similar issue, and am attempting to work out how the output `git remote -v` should translate into the filesystem hierarchy for the dev container definitions. To wit:
```
$ git remote -v
origin  ssh://git@git.server:port/project/repo.git (fetch)
origin  ssh://git@git.server:port/project/repo.git (push)
```
I've set `repositoryConfigurationPaths`, and have tried all combinations of `git@git.server:port` as the top-level directory under the repo config path: without the `git@`, without the port. I've tried using `repo` and `repo.git` as the final directory component where the `.devcontainer` directory should live, but none of these work - any attempt to rebuild/reopen the folder in the container leads to the prompt to add the development container config files afresh.

I appreciate there's no detailed doc around this (hence this issue!), but issue #3218 has a log message snippet:
```
[3840 ms] Start: Run in container: test -f /C:/Users/Connor/Documents/Github/vscode-dev-containers/repository-containers/github.com/django/django/.devcontainer/Dockerfile
```
which looks to be the path is test for a container config. Is there a way to turn this logging on to see what path my setup might be querying?

**Comment by chrmarti** [comment]
@marcussharpe The port is currently mapped to a separate path segment: `git.server/port/project/repo/.devcontainer/devcontainer.json`

To avoid that, you can move the port to an entry for that server in your `~/.ssh/config`.

I'm adding log messages to help diagnose this in the future.

**Comment by kalfa** [comment]
To be honest this ticket is about doc, not bug.

When i opened it, there was no coherent and centralised information about how the feature works.

Thanks for helping us debugging/fixing our issues, @chrmarti , but please not close the ticket changing its nature, unless the documtation has been added

**Comment by chrmarti** [comment]
We could add a note about the use of `git remote -v` and the additional log messages (once they have shipped in a stable release) to https://code.visualstudio.com/docs/remote/create-dev-container#_alternative-repository-configuration-folders.

**Comment by chrmarti** [comment]
Opened https://github.com/microsoft/vscode-docs/pull/5198 to clarify.

**Comment by marcussharpe** [comment]
@chrmarti Perfect, works a treat for me :+1: Many thanks. Looking forward to updated doc and/or log messages to facilitate this going forward.

**Comment by chrmarti** [comment]
Doc update and additional logs are in for the next stable release. (Logs are available with Remote-Containers 0.223.0 pre-release.)
